[JENKINS-9104] Add extension point that allows extensions to veto killing...

DanielWeber opened the pull request:
... of certain processes.

This is only a part of a solution for the issue, the rest should be taken care of in plugins.

This pull request replaces https://github.com/jenkinsci/jenkins/pull/1562

oleg-nenashev commented on core/src/main/java/hudson/util/ProcessKillingVeto.java:
```diff
@@ -0,0 +1,60 @@
+	 * @param p The process that is about to be killed
+	 * @return true if the process should <em>not</em> be killed, false else.
+	 */
+	public abstract boolean vetoProcessKilling(IOSProcess p);
```
I would crete a <code>VetoCause</code> class, which would describe the reason of the veto. After that, you could output this reason to logs

oleg-nenashev commented on core/src/main/java/hudson/util/ProcessTree.java:
```diff
@@ -167,6 +167,19 @@ public void killAll(Process proc, Map<String, String> modelEnvVars) throws Inter
+     * @param p The process to check
+     * @return The first extension that vetoes killing of p. null if no one objects.
+     */
+	protected ProcessKillingVeto getVeto(OSProcess p) {
```
add <code>@CheckForNull</code>

oleg-nenashev commented on core/src/main/java/hudson/util/ProcessKillingVeto.java:
```diff
@@ -0,0 +1,60 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2004-2009, Sun Microsystems, Inc., Kohsuke Kawaguchi
```
You forgot to modify the license

oleg-nenashev commented on test/src/test/java/hudson/util/ProcessTreeKillerTest.java:
```diff
@@ -55,4 +68,42 @@ public void processProperlyKilledUnix() throws Exception {
+    
+    @TestExtension("considersKillingVetos")
+    public static class VetoAllKilling extends ProcessKillingVeto {
+		@Override
```
Use spaces instead of tabs

oleg-nenashev commented on core/src/main/java/hudson/util/ProcessKillingVeto.java:
```diff
@@ -0,0 +1,60 @@
+	 */
+	public static List<ProcessKillingVeto> all() {
+		Jenkins jenkins = Jenkins.getInstance();
+		if ( null == jenkins ) return Collections.emptyList();
```
Commonly in Jenkins we use another order of parameters in comparisons. I'm not sure if it's defined somewhere in code conventions

DanielWeber commented:
@oleg-nenashev: Thanks for the quick reply. I cleaned up my whitespace issues and corrected the copyrights notice. I also added a VetoCause as you suggested. Makes sense to make it possible to analyze why a process is not being killed.

oleg-nenashev commented on core/src/main/java/hudson/util/ProcessTree.java:
```diff
@@ -455,10 +473,11 @@ public void killAll(Map<String, String> modelEnvVars) throws InterruptedExceptio
-                else
+                if (matched) { 
+                    if (getVeto(p) == null)
+                        p.killRecursively();
```
IMO logging is missing here.

oleg-nenashev commented on test/src/test/java/hudson/util/ProcessTreeKillerTest.java:
```diff
@@ -1,21 +1,34 @@
     @Rule
     public JenkinsRule j = new JenkinsRule();
-
+	private Process process;
```
Minor: tabs instead of spaces

oleg-nenashev commented on core/src/main/java/hudson/util/ProcessKillingVeto.java:
```diff
@@ -0,0 +1,102 @@
+ * See JENKINS-9104
+ * 
+ * @author Daniel Weber <daniel.weber.dev@gmail.com>
+ */
```
Minor (for merge): add <code>@since TODO</code>. Such info is very useful for plugin developers

oleg-nenashev commented:
LGTM, :+1: for merge

oleg-nenashev commented:
@DanielWeber FYI, the extension point won't work correctly for Windows platforms (see JENKINS-19156). BTW, it's not related to your PR

jglick commented on core/src/main/java/hudson/util/ProcessKillingVeto.java:
```diff
@@ -0,0 +1,102 @@
+   /**
+    * Describes the cause for a process killing veto.
+    */
+   public static class VetoCause {
```
Since you are not using this cause for any purpose that I can see, it would be return to just return a `boolean`: kill or do not kill.

oleg-nenashev replied:
@jglick I've proposed <code>VetoCause</code> in the previous review. The idea was to have a method, which would describe veto reasons in logs, etc.

jglick replied:
> The idea was to have a method, which would describe veto reasons in logs

Fine in principle, but I see no actual use of it which would demonstrate the value of the added API complexity.

oleg-nenashev commented on core/src/main/java/hudson/util/ProcessTree.java:
```diff
@@ -167,6 +169,22 @@ public void killAll(Process proc, Map<String, String> modelEnvVars) throws Inter
+        for (ProcessKillingVeto vetoExtension : ProcessKillingVeto.all()) {
+            VetoCause cause = vetoExtension.vetoProcessKilling(p);
+            if (cause != null) {
+                LOGGER.finest("Killing of pid " + p.getPid() + " vetoed by " + cause.getSource().getClass().getName() + ": " + cause.getCause());
```
IMO <code>VetoCause</code> should have a <code>getMessage()</code> (or something like that) method describing the veto reason.

jglick commented on core/src/main/java/hudson/util/ProcessKillingVeto.java:
```diff
@@ -0,0 +1,102 @@
+    *         null else.
+    */
+   @CheckForNull
+   public abstract VetoCause vetoProcessKilling(@Nonnull IOSProcess p);
```
Anyway if `VetoCause` is to be kept, it is weird to force the extension to return a struct including `this`. Should only need to return the `String` message.

DanielWeber commented:
@jglick @oleg-nenashev Let's keep the VetoCause, but remove the "source" field & ctor parameter. It will only carry a String. The VetoCause It is actually used for logging inside ProcessTree, where the veto is evaluated (getVeto()). 

Do you mind if I squash all changes into a single commit? I think we don't need that history...

oleg-nenashev commented:
@DanielWeber 
Yes, a single commit is a preferred approach in Jenkins. The proposed approach addresses all my comments. I also recommend to add <code>@since TODO</code> to new classes being introduced by your PR

DanielWeber commented:
Done. VetoCause does not longer have a "source" field, "@since TODO" is present and everything is done in a single commit.

oleg-nenashev commented:
:+1:

KostyaSha commented on core/src/main/java/hudson/util/ProcessTree.java:
```diff
@@ -167,6 +169,22 @@ public void killAll(Process proc, Map<String, String> modelEnvVars) throws Inter
+        for (ProcessKillingVeto vetoExtension : ProcessKillingVeto.all()) {
+            VetoCause cause = vetoExtension.vetoProcessKilling(p);
+            if (cause != null) {
+                LOGGER.finest("Killing of pid " + p.getPid() + " vetoed by " + vetoExtension.getClass().getName() + ": " + cause.getMessage());
```
Slow logging

oleg-nenashev commented:
@DanielWeber 
Could you update the logging?

@jglick 
Are you OK with the current <code> VetoCause</code> implementation?

DanielWeber commented:
@oleg-nenashev What's wring with the logging? It will only be executed in the rare case that someone vetoed process killing.

oleg-nenashev commented:
@DanielWeber 
I doubt it is a rare case on all installations. @KostyaSha's comment is related to the full formatting of the string before checking against the logging level. Anyway, it's possible to fix it during the merge.

oleg-nenashev commented:
@DanielWeber 
Hi Daniel, I've dig in to the code during the merge and discovered an implementation issue.

The most of calls use this method to terminate processes:
```
public void killAll(Process proc, Map<String, String> modelEnvVars) throws InterruptedException {
        LOGGER.fine("killAll: process="+proc+" and envs="+modelEnvVars);
        OSProcess p = get(proc);
        if(p!=null) p.killRecursively();
        if(modelEnvVars!=null)
            killAll(modelEnvVars);
    }
```

You check vetoes in <code>killAll(modelEnvVars)</code> overrides, so actually the process may be terminated before the check. I think the PR should be slightly reworked. The best way would be to check vetoes in <code>OSProcess::kill()</code> calls

DanielWeber commented:
@oleg-nenashev Oops, sorry. I will look into it.

olivergondza commented on test/src/test/java/hudson/util/ProcessTreeKillerTest.java:
```diff
@@ -55,4 +68,42 @@ public void processProperlyKilledUnix() throws Exception {
+        // on some platforms where we fail to list any processes, this test will
+        // just not work
+        if (ProcessTree.get() == ProcessTree.DEFAULT)
+            return;
```
This should use `Assume` to correctly skip the test.

daniel-beck commented on core/src/main/java/hudson/util/ProcessKillingVeto.java:
```diff
@@ -0,0 +1,92 @@
+        private final String message;
+
+        /**
+         * @param message A string describing the reason for the veto
```
How would this be localized?

DanielWeber replied:
Do we have to localize it? It does not show up in the UI.
How is this handled elsewhere in Jenkins? Can we assume plugins provide localized messages already?

daniel-beck replied:
Other parts in core use `org.jvnet.localizer.Localizable` in place of a string, e.g. `HealthReport`.

If it's not intended for UI use (including build logs), a string is probably fine.

daniel-beck commented:
@DanielWeber Please let me know if I can help you address the review comments. It would be great if this were finished in time for the next LTS baseline in a month or so.

DanielWeber committed:
[JENKINS-9104] Add ProcessKillingVeto extension point

This allows extensions to veto killing of certain processes.

Issue 9104 is not yet solved by this, it is only part of the solution. The
rest should be taken care of in plugins.

DanielWeber commented:
Hi all, I just updated my PR to fix the issue @oleg-nenashev found. Checking vetos is now done inside OSProcess::kill(). For systems other than Unix, OSProcess::killRecursively() also checks for vetoes since Jenkins does not traverse the process tree itself but kills complete trees at once. On Unix, the individual kill() calls done by killRecursively() handle vetoing.

=> The behavior is slightly different depending on the OS, but that's true for the whole process killing part (and subject of other Jira issues).

daniel-beck commented:
@jglick Why do you consider this to be WIP?

jglick commented:
> Why do you consider this to be WIP?

@oleg-nenashev added that to the title, I was merely adding the new label to match. (IMO we can delete the subject prefixes now that we have the new label.)

oleg-nenashev commented:
The plugin was WiP till the last patch by @DanielWeber . Now it requires a review

oleg-nenashev commented on test/src/test/java/hudson/util/ProcessTreeKillerTest.java:
```diff
@@ -1,21 +1,35 @@
+    
+    @After
+    public void tearDown() throws Exception {
+        if (null != process)
```
Probably it could be just handled in a <code>finally</code> block without external fields. Just to keep things simple

DanielWeber replied:
Can you change this while merging? I think it is not critical.

oleg-nenashev commented on core/src/main/java/hudson/util/ProcessTree.java:
```diff
@@ -228,6 +230,22 @@ void killByKiller() throws InterruptedException {
+         * @return The first non-null {@link VetoCause} provided by a process killing veto extension for this OSProcess. 
+         * null if no one objects killing the process.
+         */
+        protected @CheckForNull VetoCause getVeto() {
```
Not important: Could be public

daniel-beck replied:
Disagree, let's not add public API for no good reason.

daniel-beck commented:
I'm :+1: on this. Any other opinions?

daniel-beck commented:
@DanielWeber Do you have an implementation for this for e.g. MSBuild plugin? To make sure the use case is covered?

olivergondza commented:
:+1:

oleg-nenashev commented:
:+1:
